### ISSUE TITLE
creating a single .dll file as addon for application and integrating WebView2Loader.dll

### ISSUE THREAD

**Issue by MinesService** [MicrosoftEdge/WebView2Feedback#232]
Hello,

I am writing a Library, which is a customizing / plugin in a third-party application.  This is copied to a "Customizing" folder and loaded into the application if existing.
I included WebView2 in a C# Winforms form, but as soon as I am trying to open the form in my application I get the following error message:

> Object reference not set to an instance of an object.

May it be possible, that it has something to do with a missing WebView2Loader.dll file?

Is there any possibility to include it WebView2Loader.dll to my dll file?

**Comment by oggy22** [comment]
Thank you for reporting the issue @MinesService. It's hard to tell what is wrong without providing some more details e.g. callstack. What object is null and expected not to be null? Did you make sure you have the correct Edge version installed?

WebView2Loader.dll is part of the nuget package. I believe that if you are creating a nuget package you need to list WebView2 as dependency, anything else would be some kind of hack

**Comment by MinesService** [comment]
I am completely aware of the fact, that my description is a bit vague.
The problem for me is, that the error message does not contain any info which object reference is causing the error.

The complete error message is as follows:

````Error in easyjob Customizing
Object reference not set to an instance of an object.
Microsoft.Web.WebView2.WinForms
   at Microsoft.Web.WebView2.WinForms.WebView2.OnVisibleChanged(EventArgs e)
   at System.Windows.Forms.Control.AssignParent(Control value)
   at System.Windows.Forms.Control.ControlCollection.Add(Control value)
   at System.Windows.Forms.Form.ControlCollection.Add(Control value)
   at protonic.easyjob.Customizing.MinesService.GuiPlugin.PowerBIBrowserForm.InitializeComponent()
   at protonic.easyjob.Customizing.MinesService.GuiPlugin.PowerBIBrowserForm..ctor()
   at protonic.easyjob.Customizing.MinesService.GuiPlugin.CusApplication.MenuClick(String Name, Object Menu)
   at protonic.easyjob.Application.mdiMain.e3(Object A_0, ToolClickEventArgs A_1)
   at Infragistics.Win.UltraWinToolbars.UltraToolbarsManager.OnToolClick(ToolClickEventArgs e)
   at Infragistics.Win.UltraWinToolbars.UltraToolbarsManager.FireEvent(ToolbarEventIds id, EventArgs e)
   at Infragistics.Win.UltraWinToolbars.ToolBase.OnToolClick()
   at Infragistics.Win.UltraWinToolbars.ToolMenuItem.OnClick()
   at Infragistics.Win.UltraWinToolbars.PopupMenuItemUIElement.DoClickProcessing(MouseEventArgs e)
   at Infragistics.Win.UltraWinToolbars.PopupMenuItemUIElement.OnMouseUp(MouseEventArgs e)
   at Infragistics.Win.TextUIElementBase.OnMouseUp(MouseEventArgs e)
   at Infragistics.Win.ControlUIElementBase.ProcessMouseUpHelper(Object sender, MouseEventArgs e)
   at Infragistics.Win.ControlUIElementBase.ProcessMouseUp(Object sender, MouseEventArgs e)
   at Infragistics.Win.Utilities.ProcessEvent(Control control, ProcessEvent eventToProcess, EventArgs e)
   at Infragistics.Win.UltraWinToolbars.PopupControlBase.OnMouseUp(MouseEventArgs e)
   at System.Windows.Forms.Control.WmMouseUp(Message& m, MouseButtons button, Int32 clicks)
   at System.Windows.Forms.Control.WndProc(Message& m)
   at System.Windows.Forms.Control.ControlNativeWindow.OnMessage(Message& m)
   at System.Windows.Forms.Control.ControlNativeWindow.WndProc(Message& m)
   at System.Windows.Forms.NativeWindow.Callback(IntPtr hWnd, Int32 msg, IntPtr wparam, IntPtr lparam) ````

The software I am writing the customizing / plugin for has the name "easyjob".

I just need a solution to add a single .dll-file to the "Customizing" folder, which gives me the possibility to open a modern website using several scripts like Power BI within a form.

**Comment by pagoe-msft** [comment]
@MinesService 

Could you try following the first few steps of the WinForms Getting Started Guide:
https://docs.microsoft.com/en-us/microsoft-edge/webview2/gettingstarted/winforms

and let me know if you get the same error?

If not, then it is likely how your product is set up. Have you included the correct NuGet package? It is 0.9.511-prerelease.

**Comment by shawty** [comment]
@MinesService what you might be experiencing is a similar issue to the one I encountered.

When you first instantiate the WebView control, it takes it a few seconds before the internal "CoreWebView2" property is initialized.

When I first started using the control I was doing things like this

```
public FrmMainForm()
{
  InitializeComponent();
  webView.CoreWebView2Ready += WebView_CoreWebView2Ready;
  webView.CoreWebView2.WebMessageReceived += CoreWebView2_WebMessageReceived;
}
```

In my windows forms constructor.

This kept throwing a null reference which took me a while to track down.

When I eventually figured it out, I found that it was throwing the exception in the last line where I'm setting up the WebMessageRecived handler, and it was because the "CoreWebView2" property was null.

This is due to it taking time to initialize the edge browser and load things.

If you look at my small code snippet, you'll see I define 2 handlers, one for the web message and one for the control ready.  The solution is to move any code that relies on the "CoreWebView2" property being NOT null into a handler that fires once the control is ready, for example:

```
public FrmMainForm()
{
  InitializeComponent();
  webView.CoreWebView2Ready += WebView_CoreWebView2Ready;
}

private void WebView_CoreWebView2Ready(object sender, EventArgs e)
{
  webView.CoreWebView2.WebMessageReceived += CoreWebView2_WebMessageReceived;
}
```

If you wait until you get the core web view ready event, at that point you can be sure the property will have been initialized and you can continue to populate your environment with your custom code and calls.
